### ISSUE TITLE
Oxygen production build fails - TypeError: Cannot read properties of null (reading 'parse')

### ISSUE THREAD

**Issue by sebinievas** [BuilderIO/builder-shopify-hydrogen#3]
When attempting to deploy to Oxygen, the system fails due to the following error:

`Cannot read properties of null (reading 'parse')`

I've installed a vanilla copy of hydrogen and I started copying over code line by line. When I included the builder client component, the `builder.init` method causes this error to be thrown. When I run hydrogen in dev mode, it renders just fine. But when it builds for production (as it needs to in order to deploy to Oxygen) the script cannot proceed past that point so it cannot complete a successful deployment.

**Comment by steve8708** [comment]
Thanks for reporting @sebinievas - we hadn't tested deploying to oxygen just yet, but we can test that out this week and should be able to get a fix in shortly, stay tuned!

**Comment by sebinievas** [comment]
> Thanks for reporting @sebinievas - we hadn't tested deploying to oxygen just yet, but we can test that out this week and should be able to get a fix in shortly, stay tuned!

No problem. Any news on this? We can't go live without it, and we'll have to use something else unfortunately if not. Thanks again for fixing this.

**Comment by steve8708** [comment]
Absolutely - @Dkendal was able to reproduce the issue and is working on a fix for this as we speak

**Comment by sebinievas** [comment]
I'm not sure if this helps but I did find this issue that was reported on Vite, but it ended up being the vendor that ultimately resolved it. Perhaps it contains some insights. https://github.com/vitejs/vite/issues/8395

**Comment by Dkendal** [comment]
Hey @sebinievas, sorry for the radio silence! I have a fix that I'm pushing today.

There will be a few fixes to the @builder.io/sdk that will fix the issue that you reported above: removing the usage of `require('url').parse` which isn't available in the miniflare env.

You will also have to include a vite plugin that I'll be publishing that monkey patches out code generation apis in the SSR flow. I'll comment with other instructions once it's published.

**Comment by sebinievas** [comment]
This is super great to hear, thanks for working on this. Just something to note, I believe hydrogen is currently using `vite ^2.9.0`. I know they are committed to vite and are working on getting 3.0 to work. Just something to keep on the radar, hopefully this works for both versions. Thanks again @Dkendal and @steve8708 !

**Comment by Dkendal** [comment]
For transparency and posterity for future readers: both issues described above were due to differences in the nodejs (development) and miniflare/minioxygen (preview and production) environments.

Oxygen runs on Cloudflare workers, and the Oxygen preview runs on Miniflare.

Oxygen's docs say that it requires nodejs >= 16.5.  There is an additional section that describes it's "runtime api[^ox]" which does omit a number of functions and modules - it appears that they don't include deprecated nodejs apis. For the builder sdk we haven't replaced all usages of `require('url').parse` to the `URL` constructor - I have a PR that will do this that I mentioned above.

For the Builder SDK to handle custom, user defined code that was defined in the editor we rely on code generation and eval apis, specifically: `eval`, `new Function`, and the `vm` module. The oxygen runtime docs [^ox] have a specifc call out that these API's are not available. What this means for us is that SSR functionality that relies on this will be disabled. I have a vite plugin that accomplishes this by monkey patching out eval calls in the @builder/react and @builder/sdk packages in the current project. In my testing I wasn't actually able to get these code paths to trigger so if you encounter problems just let me know and I can look into it.

Assuming that vite 3 uses compatable plugins, which are really just rollup plugins this should still work.

We have a new version of the react sdk in progress at the moment, we'll be looking at producing seperate builds for client and server code there that should handle this more elegantly.

[^ox]: https://shopify.dev/custom-storefronts/oxygen/worker-runtime-apis
[^cf]: https://developers.cloudflare.com/workers/runtime-apis

**Comment by Dkendal** [comment]
Alright the changes are up for review now, you can sub to the PR here: https://github.com/BuilderIO/builder/pull/1184. I'll publish once it's reviewed.

Linked in the PR is the [demo project](https://github.com/Dkendal/builder-shopify-hydrogen/blob/v2/vite.config.ts) I'm building against - a lot has changed in the Hydrogen npm package by the looks for it since we initially published this repo 6 months so it's just a fresh Hydrogen starter with the /builder/** path mapped to use a custom builder component.

**Comment by Dkendal** [comment]
Hey @sebinievas, my PR is still under review but I could a dev release that will hopefully unblock you.

Try adding `"@builder.io/react": "2.0.2-9"` to your project.

Let me know if it works out for you!

**Comment by sebinievas** [comment]
@Dkendal I tried it, but I'm still seeing the parse error. To get this behavior, use the latest version of hydrogen and add the package as you mentioned. Then use `yarn build` followed by `yarn preview` to emulate Oxygen's behavior. This is the only way i've been able to replicate the behavior when attempting to deploy to Oxygen land.

**Comment by Dkendal** [comment]
Hey @sebinievas, I'll try with a fresh starter and see if I get the same issue. Would you be able to create a repo that demonstrates the issue so I can eliminate any other variables? I'd also be open to jumping on a video call to see if I can work through this with you.

**Comment by Dkendal** [comment]
It just occured to me as well that you'd have to make sure that @builder.io/sdk is being upgraded to the `dev` tag, which should be `1.1.27-5` - this should happen automatically but check and report back what version it's pinned to in your lockfile just to be sure.

**Comment by sebinievas** [comment]
> It just occured to me as well that you'd have to make sure that @builder.io/sdk is being upgraded to the `dev` tag, which should be `1.1.27-5` - this should happen automatically but check and report back what version it's pinned to in your lockfile just to be sure.

Hmm, I just re-installed it and I am showing the correct versions. Attaching screenshot just in case.

<img width="1264" alt="image" src="https://user-images.githubusercontent.com/382021/184141460-23fb64fa-ce7f-4602-a8e9-c3cc25f191fe.png">

**Comment by sebinievas** [comment]
@Dkendal OK, I removed the rollup-plugin stuff I was attempting to use and it seems to be working now! Hooray! 🎊

**Comment by Dkendal** [comment]
Woohoo! Good to hear! No you don't need the rollup plugin - it ended up not being required I should have mentioned that in the last message.

There's a very good chance that you'll hit one or two other speed bumps after you start using data bindings as I wasn't able to fully validate that at the moment - do not hestitate to reach out in this thread, a new issue, or on my email dylan@builder.io I wanna make sure that your team has as smooth as possible of a process with oxygen, it's new territory for us as well.
